### PR TITLE
Navigator: add terrain collision avoidance during descents

### PR DESCRIPTION
**Pull request by sfuhrer**
Alternative to https://github.com/PX4/PX4-Autopilot/pull/23396. Changes:
- do not exclusively apply terrain collision logic for VTOL in landing phase
- do not allow to increase the altitude setpoint above the current altitude
- implement user feedback

### Solved Problem
There are several reasons why during a landing approach (landing with a pattern) a VTOL might fly into terrain.
The baro bias estimate could not be correct (especially when flying GPS denied) or the user might have not planned the mission correctly.

### Solution
Avoid flying into terrain using the distance sensor. Enable through the parameter `NAV_MIN_GND_DIST`.
Only active during commanded descents with vz>0 (to prevent climb-aways), excluding landing and VTOL transitions.
It changes the altitude setpoint in the triplet to maintain the current altitude and republish the triplet.
We also change the mission item altitude used for acceptance calculations to prevent getting stuck in a loop.

### Changelog Entry
For release notes: 
```
Feature: Navigator: add terrain collision avoidance during descents
New parameter: NAV_MIN_GND_DIST
Documentation: todo
```

### Alternatives
Have it more exclusively for VTOL landings: https://github.com/PX4/PX4-Autopilot/pull/23396, though I think this feature would also be helpful for e.g. plane landings (trigger the exit of the loiter down before landing based on distance sensor).

### Test coverage
SITL tested with VTOLs, planes and MCs, all RTL types and mission landings.

### CODE REVIEW

File: src/modules/navigator/mission_block.cpp
Comment (MaEtUgR) [inline src/modules/navigator/mission_block.cpp:1022]:
```suggestion
	static constexpr float kAltitudeDifferenceForDescentCondition = 5.f; // meters threshold
```
So from 4 meters altitude it's still fine to fly into the ground? This is probably only there to make sure it doesn't trigger (again) if the position estimate is 10cm higher than the triplet (that was already adjusted) 🤔

**Review by MaEtUgR — COMMENTED** [review]
I think the idea makes sense but the implementation is sadly pretty hacky, I hope it doesn't create too many side effects.

The worst one I could currently think off is if you are operating a multicopter, the distance sensor is blocked by the payload, it doesn't descend anymore e.g. could fly into a bridge if the mission would go underneath. Also would RTL still be usable in that case because the phase where you're descending but not landing yet wouldn't descend right?

**Comment by sfuhrer** [comment]
>Also would RTL still be usable in that case because the phase where you're descending but not landing yet wouldn't descend right?

Yes correct, I just tested it, it would have the same behavior for MC RTL as for VTOL/Plane RTL: while it is descending (for MC it would be up to RTL_DESCEND_ALT) the terrain collision checks are active. If during this time the distance sensor measurement is wrong (e.g. 1m), it would trigger the "Terrain collision risk" warning and adapt the current altitude setpoint to the current altitude. Once above the landing point it will then still set the setpoint type to Land and come down, so beside the warning the user wouldn't notice anything else.

File: src/modules/navigator/mission_block.cpp
Comment (sfuhrer) [reply src/modules/navigator/mission_block.cpp]: Yes correct, that was the reason I put it there, and yes, that means if the vehicle is less than 5m vertical distance away from the setpoint it wouldn't trigger the collision prevention. Very random number, I agree. 
I've reduced it to 2m to mitigate it slightly. Otherwise, to get rid of the magic threshold, we could only allow the terrain correction once per set triplet (and reset if a new triplet comes in).

**Review by MaEtUgR — APPROVED** [review]
As a workaround that's acceptable. I'd try to find a better solution long term.

**Pull request merged**
